File: backend/src/main/java/com/gearline/marketplace/common/util/ShippingCalculator.java
package com.gearline.marketplace.common.util;

import com.gearline.domain.product.Dimensions;
import com.gearline.domain.product.Product;
import com.gearline.marketplace.common.dto.ShippingDetails;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Map;

/**
 * Computes shipping parameters from product physical data and per-listing overrides.
 *
 * Unit conversions:
 *   Weight: kilograms → ounces  (1 kg = 35.27396 oz)
 *   Length: centimetres → inches (1 cm = 0.393701 in)
 *
 * Insurance tiers:
 *   The declared package value is rounded up to the nearest $1,000 to match
 *   the tier pricing used by most shipping carriers and third-party insurer
 *   integrations (e.g. Shipstation, EasyPost). A $2,400 guitar becomes a
 *   $3,000 declared value; a $1,000 guitar stays at $1,000.
 */
@Component
@Slf4j
public class ShippingCalculator {

    private static final BigDecimal KG_TO_OZ    = new BigDecimal("35.27396");
    private static final BigDecimal CM_TO_IN     = new BigDecimal("0.393701");
    private static final BigDecimal INSURANCE_INCREMENT = new BigDecimal("1000");
    private static final int        SCALE        = 3;

    // ── Unit conversions ───────────────────────────────────────────────────────

    public BigDecimal kgToOz(BigDecimal kg) {
        if (kg == null) return null;
        return kg.multiply(KG_TO_OZ).setScale(SCALE, RoundingMode.HALF_UP);
    }

    public BigDecimal cmToIn(BigDecimal cm) {
        if (cm == null) return null;
        return cm.multiply(CM_TO_IN).setScale(SCALE, RoundingMode.HALF_UP);
    }

    // ── Insurance calculation ─────────────────────────────────────────────────

    /**
     * Rounds a declared item value up to the nearest $1,000 insurance tier.
     *
     * Examples:
     *   $999   → $1,000
     *   $1,000 → $1,000
     *   $1,001 → $2,000
     *   $2,400 → $3,000
     *   $5,000 → $5,000
     */
    public BigDecimal calculateInsuranceValue(BigDecimal price) {
        if (price == null || price.compareTo(BigDecimal.ZERO) <= 0) {
            return BigDecimal.ZERO;
        }
        // ceiling division: ⌈price / 1000⌉ * 1000
        BigDecimal tiers = price.divide(INSURANCE_INCREMENT, 0, RoundingMode.CEILING);
        return tiers.multiply(INSURANCE_INCREMENT);
    }

    // ── Composite resolver ─────────────────────────────────────────────────────

    /**
     * Resolves a complete ShippingDetails from a product's physical data plus any
     * per-listing overrides stored in {@code listing_overrides}.
     *
     * Override keys consumed from the overrides map:
     * <pre>
     *   reverb_shipping_profile_name  — Reverb seller shipping profile ID (numeric string)
     *   weight_oz_override            — explicit imperial weight in oz (skips kg→oz conversion)
     * </pre>
     * eBay-specific keys (ebay_fulfillment_policy_id, etc.) are NOT consumed here —
     * they pass through to extraParams and are read directly by EbayConnector.
     *
     * @param product   canonical product (provides weight_kg, dimensions in cm, price)
     * @param overrides listing_overrides JSONB map; may be empty but never null
     */
    public ShippingDetails resolveShipping(Product product, Map<String, Object> overrides) {
        // Weight: check for an explicit imperial override first, else convert from kg
        BigDecimal weightOz = getDecimalOverride(overrides, "weight_oz_override");
        if (weightOz == null) {
            weightOz = kgToOz(product.getWeightKg());
        }

        // Dimensions: stored in inches directly (entered by seller in Shopify as inches)
        BigDecimal lengthIn = null;
        BigDecimal widthIn  = null;
        BigDecimal heightIn = null;
        Dimensions dims = product.getDimensions();
        if (dims != null) {
            lengthIn = dims.getLengthIn();
            widthIn  = dims.getWidthIn();
            heightIn = dims.getHeightIn();
        }

        // Reverb shipping profile ID — Reverb-only concept.
        // eBay uses a fulfillment policy UUID sent separately via extraParams, not here.
        String shippingProfileName = getStringOverride(overrides, "reverb_shipping_profile_name");

        // Insurance value
        BigDecimal insuranceValue = calculateInsuranceValue(product.getPrice());

        if (weightOz == null && lengthIn == null && shippingProfileName == null) {
            log.debug("Product {} has no weight, dimensions, or shipping profile — shipping block will be omitted",
                product.getSku());
        }

        return ShippingDetails.builder()
            .weightOz(weightOz)
            .lengthIn(lengthIn)
            .widthIn(widthIn)
            .heightIn(heightIn)
            .shippingProfileName(shippingProfileName)
            .insuranceValueUsd(insuranceValue)
            .build();
    }

    // ── Helpers ────────────────────────────────────────────────────────────────

    private String getStringOverride(Map<String, Object> overrides, String key) {
        Object val = overrides.get(key);
        return (val instanceof String s) ? s : null;
    }

    private BigDecimal getDecimalOverride(Map<String, Object> overrides, String key) {
        Object val = overrides.get(key);
        if (val == null) return null;
        try {
            return new BigDecimal(val.toString());
        } catch (NumberFormatException e) {
            log.warn("Invalid decimal override for key '{}': {}", key, val);
            return null;
        }
    }
}

File: backend-node/src/services/shipping-calculator.ts
import type { ProductRow } from '../db/types.js';
import {
  ceilingToMultiple,
  decimalToString,
  multiplyDecimal,
  parseDecimal,
  setScaleHalfUp,
  tryParseDecimal,
} from '../util/decimal.js';

/**
 * Port of ShippingCalculator.
 *
 * All arithmetic runs through the BigInt decimal helpers rather than JS numbers.
 * These values are sent to eBay and Reverb as package weights and declared
 * insurance values, so a float rounding artefact would show up as a wrong
 * shipping quote on a live listing.
 */

const KG_TO_OZ = parseDecimal('35.27396');
const INSURANCE_INCREMENT = 1000n;
const SCALE = 3;

export interface ShippingDetails {
  /** Package weight in ounces. */
  weightOz: string | null;
  /** Package dimensions in inches. */
  lengthIn: string | null;
  widthIn: string | null;
  heightIn: string | null;
  /** Reverb seller shipping profile ID. eBay uses a policy ID via extraParams. */
  shippingProfileName: string | null;
  /** Declared value, rounded up to the nearest $1,000 tier. */
  insuranceValueUsd: string | null;
}

export function kgToOz(kg: string | null): string | null {
  const parsed = tryParseDecimal(kg);
  if (!parsed) return null;
  return decimalToString(setScaleHalfUp(multiplyDecimal(parsed, KG_TO_OZ), SCALE));
}

/**
 * Rounds a declared value UP to the next $1,000 tier.
 *
 *   $999   → $1,000
 *   $1,000 → $1,000   (exact multiples do not jump a tier)
 *   $1,001 → $2,000
 *   $2,400 → $3,000
 *
 * Note the exact-multiple case: ceiling division must not push $1,000 to $2,000.
 */
export function calculateInsuranceValue(price: string | null): string {
  const parsed = tryParseDecimal(price);
  if (!parsed || parsed.unscaled <= 0n) return '0';
  return decimalToString(ceilingToMultiple(parsed, INSURANCE_INCREMENT));
}

/**
 * Resolves shipping parameters from product physical data plus listing overrides.
 *
 * Override keys consumed here:
 *   weight_oz_override            — explicit imperial weight, skips kg→oz
 *   reverb_shipping_profile_name  — Reverb seller profile ID
 *
 * eBay keys are deliberately NOT consumed — they pass through to extraParams and
 * are read directly by the eBay connector.
 */
export function resolveShipping(
  product: ProductRow,
  overrides: Record<string, unknown>,
): ShippingDetails {
  const weightOverride = tryParseDecimal(overrides['weight_oz_override']);
  const weightOz = weightOverride
    ? decimalToString(weightOverride)
    : kgToOz(product.weight_kg);

  const profileRaw = overrides['reverb_shipping_profile_name'];
  const shippingProfileName = typeof profileRaw === 'string' ? profileRaw : null;

  return {
    weightOz,
    // Dimensions are stored in inches already (V16 renamed the _cm columns);
    // no conversion, despite the class name suggesting otherwise.
    lengthIn: product.dim_length_in,
    widthIn: product.dim_width_in,
    heightIn: product.dim_height_in,
    shippingProfileName,
    insuranceValueUsd: calculateInsuranceValue(product.price),
  };
}
